Update docstrings and docs (#232)
histolab/histolab: docs/api/filters.rst

   filters/morphological_filters
   filters/morphological_filters_functional

.. toctree::

> Filters in ``histolab`` are designed to be applied singularly or combined in a chain of transformations. A composition of filters is predefined for tissue segmentation, while custom filter combinations can be used for tissue detection or other tasks.

Reviewer: all this should go before "API reference" and the list of submodules

> * Image filters:

Reviewer: this is useful but too wall text, we should find a way to present the same information in a cleaner way.. maybe we could add images?

Author: Do u have suggestions? anyway I agree with images, we have plenty of them

To ease combining filters together, inspired by the ``transforms`` module of torchvision [11]_, we implemented the ``Compose`` class in the ``filters`` subpackage. ``Compose`` allows ``histolab`` filters - both image and morphological filters - to be concatenated as a chain of functions, without any intermediate transformation or prior knowledge on their ordering. In order to enable this composition, filters have been designed to follow the same usage pattern, and to input/output either a PIL Image or a NumPy array object.
To clarify how the ``Compose`` object works, let us consider the sequential application of a set of filters sequentially, where the output of a filter is the input of the following one:

.. code-block:: python3

>         from PIL import Image
>         from histolab.filters.image_filters import RgbToGrayscale, OtsuThreshold, ApplyMaskImage
>         from histolab.filters.morphological_filters import BinaryDilation
>         def not_composed_filters(image_rgb):
>             rgb_to_grayscale = RgbToGrayscale()
>             otsu_threshold = OtsuThreshold()
>             binary_dilation = BinaryDilation()
>             apply_mask_image = ApplyMaskImage(image_rgb) # apply the resulting mask on the original image
> 
>             image_gray = rgb_to_grayscale(image_rgb)
>             image_thresholded = otsu_threshold(image_gray)
>             image_dilated = binary_dilation(image_thresholded)
>             return apply_mask_image(image_dilated)
>         image_rgb = Image.open("path/to/image.png")
>         resulting_image = not_composed_filters(image_rgb)

Reviewer: ```suggestion
    from PIL import Image

    from histolab.filters.image_filters import ApplyMaskImage, OtsuThreshold, RgbToGrayscale
    from histolab.filters.morphological_filters import BinaryDilation

    def not_composed_filters(image_rgb):
        rgb_to_grayscale = RgbToGrayscale()
        otsu_threshold = OtsuThreshold()
        binary_dilation = BinaryDilation()
        apply_mask_image = ApplyMaskImage(
            image_rgb
        )  # apply the resulting mask on the original image

        image_gray = rgb_to_grayscale(image_rgb)
        image_thresholded = otsu_threshold(image_gray)
        image_dilated = binary_dilation(image_thresholded)
        return apply_mask_image(image_dilated)

    image_rgb = Image.open("path/to/image.png")
    resulting_image = not_composed_filters(image_rgb)

```

Despite being formally correct, the above implementation in neither intuitive or economic (in terms of memory used and lines of code). The use of a ``Compose`` object leads to a more compact implementation:

.. code-block:: python3

>     from PIL import Image
>     from histolab.filters.image_filters import RgbToGrayscale, OtsuThreshold, ApplyMaskImage, Compose
>     from histolab.filters.morphological_filters import BinaryDilation
> 
>     def composed_filters(image_rgb):
>         filters = Compose([
>             RgbToGrayscale(),
>             OtsuThreshold(),
>             BinaryDilation(),
>             ApplyMaskImage(image_rgb),
>         ])
>         return filters(image_rgb)
> 
>     image_rgb = Image.open("path/to/image.png")
>     resulting_image = composed_filters(image_rgb)

Reviewer: ```suggestion
    from PIL import Image

    from histolab.filters.image_filters import (
        ApplyMaskImage,
        Compose,
        OtsuThreshold,
        RgbToGrayscale,
    )
    from histolab.filters.morphological_filters import BinaryDilation

    def composed_filters(image_rgb):
        filters = Compose(
            [
                RgbToGrayscale(),
                OtsuThreshold(),
                BinaryDilation(),
                ApplyMaskImage(image_rgb),
            ]
        )
        return filters(image_rgb)

    image_rgb = Image.open("path/to/image.png")
    resulting_image = composed_filters(image_rgb)
```

.. note::
    Although ``not_composed_filters`` and ``composed_filters`` functions return the same result, the use of ``Compose`` avoids storing intermediate results and wasting memory in case of very large input image. For example, the peak RAM usage for an image of size 19,394px x 6,136px (84.9 MB) is ~2600 MB when using ``composed_filters`` in contrast to the ~3200 MB allocated for the ``not_composed_filters`` function.

> Largest Tissue Box extraction

Reviewer: I would move this somewhere in the tiler module

Author: tiler?

Author: oh the biggest box you mean

Reviewer: yep

.. code-block:: python3

    from histolab.filters.image_filters import Compose, OtsuThreshold, RgbToGrayscale
    from histolab.filters.morphological_filters import BinaryDilation, RemoveSmallHoles, RemoveSmallObjects
    filters = Compose([
                RgbToGrayscale(),
                OtsuThreshold(),
                BinaryDilation(),
                RemoveSmallHoles(),
                RemoveSmallObjects()
        ])

To speed up the tiling procedure, the extraction area is further restricted to the bounding box including the largest connected component: only the two pairs of coordinates defining the box are then needed in the tile cropping procedure. The largest tissue box is computed in four steps: (i) for each of the connected regions of the tissue mask, a numeric label is assigned (starting from 1); (ii) a set of properties is calculated for each region (e.g.~area of the region, coordinates of the bounding box, area of the bounding box); (iii) only the coordinates of the bounding box including the region with the largest area are considered; (iii) a new binary mask is created, with True values inside of the bounding box, False outside.

> Future releases of ``histolab`` will support for the extraction of tiles from secondary regions other than the largest tissue box, such as the top *n* boxes (sorted by tissue area) or areas within the boundaries of the segmented tissue. Moreover, custom binary masks for tile extraction will be supported.

Reviewer: I would remove this

References
----------

.. [1] A Ruifrok and et al. “Quantification of histochemical staining by color deconvolution”. *Anal Quant Cytol Histol* 23.4 (2001)

> .. [2] K B Suliman and A Krzyzak. “Computerized Counting-Based System for Acute Lym-phoblastic Leukemia Detection in Microscopic Blood Images”. Artificial Neural Net-works and Machine Learning (ICANN 2018). *Springer* (2018)

Reviewer: ```suggestion
.. [2] K B Suliman and A Krzyzak. “Computerized Counting-Based System for Acute Lymphoblastic Leukemia Detection in Microscopic Blood Images”. Artificial Neural Networks and Machine Learning (ICANN 2018). *Springer* (2018)
```

.. [3] R D Peng. “Reproducible research in computational science”. Science 334.6060 (2011)
.. [4] J Sieren andet al. “An automated segmentation approach for highlighting the histologicalcomplexity of human lung cancer”. *Ann Biomed Eng* 38.12 (2010)
.. [5] E A Wagar and et al. “Specimen labeling errors: a Q-probes analysis of 147 clinical laboratories”. *Arch Pathol Lab Med* 132.10 (2008)

> .. [6] S K Phan and et al. “Biomedical Imaging Informatics for Diagnostic Imaging MarkerSelection”. *Health Informatics Data Analysis*. Springer (2017)

Reviewer: ```suggestion
.. [6] S K Phan and et al. “Biomedical Imaging Informatics for Diagnostic Imaging Marker Selection”. *Health Informatics Data Analysis*. Springer (2017)
```

> .. [7] S Ali and et al. “Ink removal from histopathology whole slide images by combining classification, detection and image generation models”. *IEEE 16th International Symposiumon Biomedical Imaging* (ISBI 2019).

Reviewer: ```suggestion
.. [7] S Ali and et al. “Ink removal from histopathology whole slide images by combining classification, detection and image generation models”. *IEEE 16th International Symposium on Biomedical Imaging* (ISBI 2019).
```

.. [8] M Dusenberry and et al. `deep-histopath <https://github.com/CODAIT/deep-histopath>`

> .. [9] A P Vartak and V Mankar. “Morphological image segmentation analysis”. *Int J Com-put Appl* 6.2 (2013)

Reviewer: ```suggestion
.. [9] A P Vartak and V Mankar. “Morphological image segmentation analysis”. *Int J Comput Appl* 6.2 (2013)
```

.. [10] L Vincent and P Soille. “Watersheds in digital spaces: an efficient algorithm based on immersion simulations”. *IEEE Trans Pattern Anal Mach Intell* 6 (1991)

> .. [11] A Paszke and et al. “PyTorch: An Imperative Style, High-Performance Deep LearningLibrary”. *Advances in Neural Information Processing Systems* 32 (NeurIPS  2019)

Reviewer: ```suggestion
.. [11] A Paszke and et al. “PyTorch: An Imperative Style, High-Performance Deep Learning Library”. *Advances in Neural Information Processing Systems* 32 (NeurIPS  2019)
```

> .. note:: ``histolab`` stores masks as NumPy arrays. The utility class ``ToPILImage`` in the image filters module retrieves the Pillow Image from the corresponding array.

Reviewer: please add link to `ToPILImage`

* Morphological filters:

>     * **Image preprocessing**: Morphology is a comprehensive set of image processing operations that transform images by using geometrical structures. Morphological operations act by adjusting image pixels based on the value of the pixels in the neighborhood. The choice of the neighborhood's size and shape will affect the behaviour of the morphological operation so that it will be sensitive to specific shapes in the input image [9]_. The *structuring element* is the component of the morphological operations that defines the considered neighborhood: it is a shape (typically a circle or a square) that determines the area used to process each pixel in the image. Usually, the shape and size of the structuring element are chosen to reflect the geometry of the objects in the image that the structuring element will transform: for example, linear structuring element would be used to detect lines in an image. Morphological operations can be applied to binary masks to shrink or enlarge regions of the image. Classic morphological operations include dilation, erosion, opening and closing; ``histolab`` implements these operations in the filters submodule ``morphological_filters``. The default structuring element is a disk, with radius 5 for both dilation and erosion, and radius 3 for both opening and closing. However, it is possible to override the default value by passing ``disk_size=N`` as parameter to the filter constructor. The ``morphological_filters`` module implements three additional morphological operations useful for manipulating binary masks: `WhiteTopHat <filters/morphological_filters.html#src.histolab.filters.morphological_filters.WhiteTopHat>`_, `RemoveSmallObjects <filters/morphological_filters.html#src.histolab.filters.morphological_filters.RemoveSmallObjects>`_, and `RemoveSmallHoles <filters/morphological_filters.html#src.histolab.filters.morphological_filters.RemoveSmallHoles>`_. The *white top-hat* transformation is defined as the difference between the image and its morphological opening with respect to a structuring element. This operation results in an image including only structures smaller than the structuring element and brighter than their neighborhood, and it is thus used to extract light details on a dark background. The white top-hat filter uses a cross-shaped structuring element with connectivity 1 by default. The ``RemoveSmallObjects`` filter removes objects with an area smaller than a specified value while the ``RemoveSmallHoles`` filter "fills" holes with an area smaller than the specified threshold. The minimal area value is set to 3000 for both filters.

Reviewer: please add link to "morphological_filters", "RemoveSmallObjects", and "RemoveSmallHoles"

To ease combining filters together, inspired by the ``transforms`` module of torchvision [11]_, we implemented the `Compose <filters/image_filters.html#src.histolab.filters.image_filters.Compose>`_ class in the ``filters`` subpackage. `Compose <filters/image_filters.html#src.histolab.filters.image_filters.Compose>`_ allows ``histolab`` filters - both image and morphological filters - to be concatenated as a chain of functions, without any intermediate transformation or prior knowledge on their ordering. In order to enable this composition, filters have been designed to follow the same usage pattern, and to input/output either a PIL Image or a NumPy array object.
To clarify how the `Compose <filters/image_filters.html#src.histolab.filters.image_filters.Compose>`_ object works, let us consider the sequential application of a set of filters sequentially, where the output of a filter is the input of the following one:

.. code-block:: python3

    from PIL import Image
    from histolab.filters.image_filters import (
        ApplyMaskImage,
        OtsuThreshold,
        RgbToGrayscale,
    )
    from histolab.filters.morphological_filters import BinaryDilation

>     def not_composed_filters(image_rgb):

Reviewer: ```suggestion

    def not_composed_filters(image_rgb):
```

        rgb_to_grayscale = RgbToGrayscale()
        otsu_threshold = OtsuThreshold()
        binary_dilation = BinaryDilation()
        apply_mask_image = ApplyMaskImage(
            image_rgb
        )  # apply the resulting mask on the original image
        image_gray = rgb_to_grayscale(image_rgb)
        image_thresholded = otsu_threshold(image_gray)
        image_dilated = binary_dilation(image_thresholded)

>         return apply_mask_image(image_dilated)

Reviewer: ```suggestion
        return apply_mask_image(image_dilated)
        
```

    image_rgb = Image.open("path/to/image.png")
    resulting_image = not_composed_filters(image_rgb)

Despite being formally correct, the above implementation in neither intuitive or economic (in terms of memory used and lines of code). The use of a `Compose <filters/image_filters.html#src.histolab.filters.image_filters.Compose>`_ object leads to a more compact implementation:

.. code-block:: python3

    from PIL import Image
    from histolab.filters.image_filters import (
        ApplyMaskImage,
        Compose,
        OtsuThreshold,
        RgbToGrayscale,
    )
    from histolab.filters.morphological_filters import BinaryDilation

>     def composed_filters(image_rgb):

Reviewer: ```suggestion

    def composed_filters(image_rgb):
```

            ]
        )
        return filters(image_rgb)

    image_rgb = Image.open("path/to/image.png")
    resulting_image = composed_filters(image_rgb)

.. note::
    Although ``not_composed_filters`` and ``composed_filters`` functions return the same result, the use of `Compose <filters/image_filters.html#src.histolab.filters.image_filters.Compose>`_ avoids storing intermediate results and wasting memory in case of very large input image. For example, the peak RAM usage for an image of size 19,394px x 6,136px (84.9 MB) is ~2600 MB when using ``composed_filters`` in contrast to the ~3200 MB allocated for the ``not_composed_filters`` function.

Largest Tissue Box extraction
-----------------------------

As an application of `Compose <filters/image_filters.html#src.histolab.filters.image_filters.Compose>`_, ``histolab`` provides the automatic segmentation of tissue in a WSI by using a pre-defined composition of filters. The default chain of filters for tissue segmentation is empirically determined and it includes: conversion to grayscale, Otsu thresholding, binary dilation, small holes and small objects filtering, similarly to the pipelines in [8]_:

.. code-block:: python3

>     from histolab.filters.image_filters import Compose, OtsuThreshold, RgbToGrayscale
>     from histolab.filters.morphological_filters import BinaryDilation, RemoveSmallHoles, RemoveSmallObjects
>     filters = Compose([
>                 RgbToGrayscale(),
>                 OtsuThreshold(),
>                 BinaryDilation(),
>                 RemoveSmallHoles(),
>                 RemoveSmallObjects()
>         ])

Reviewer: ```suggestion
    from histolab.filters.image_filters import Compose, OtsuThreshold, RgbToGrayscale
    from histolab.filters.morphological_filters import (
        BinaryDilation,
        RemoveSmallHoles,
        RemoveSmallObjects,
    )

    filters = Compose(
        [
            RgbToGrayscale(),
            OtsuThreshold(),
            BinaryDilation(),
            RemoveSmallHoles(),
            RemoveSmallObjects(),
        ]
    )

```
